File: comfyui/ui/src/schemas/settings/rate-limits-schema.ts
/**
 * Rate Limits Schema
 *
 * Zod schemas for validating rate limit data stored in localStorage/sessionStorage.
 * Low priority - only affects rate limiting, not security critical.
 * Includes comprehensive validation constraints for security and data integrity.
 */

import { z } from "zod";
import {
  MAX_NAME_LENGTH,
} from "../shared-validation";

// ============================================================================
// Primitives
// ============================================================================

const nonNegativeInt = z.number().int().nonnegative();
const positiveFinite = z.number().finite().positive();

// ============================================================================
// Rate Limit Config Schema
// ============================================================================

export const RateLimitConfigSchema = z.object({
  toolName: z.string().min(1).max(MAX_NAME_LENGTH).trim(),
  maxPerDay: positiveFinite.max(100000), // Max 100k calls per day
  maxPerSession: positiveFinite.max(10000).optional(), // Max 10k calls per session
  vramCostMB: positiveFinite.max(1000000).optional(), // Max 1TB VRAM cost
}).strict();

export type RateLimitConfig = z.infer<typeof RateLimitConfigSchema>;

// ============================================================================
// Rate Limit Status Schema
// ============================================================================

export const RateLimitStatusSchema = z.object({
  toolName: z.string().min(1).max(MAX_NAME_LENGTH).trim(),
  callsToday: nonNegativeInt.max(100000), // Max 100k calls today
  maxPerDay: positiveFinite.max(100000), // Max 100k calls per day
  callsThisSession: nonNegativeInt.max(10000), // Max 10k calls this session
  maxPerSession: positiveFinite.max(10000).optional(), // Max 10k calls per session
  canCall: z.boolean(),
  blockedReason: z.string().max(500).trim().optional(),
  resetsAt: z.string().datetime(), // ISO timestamp
  resetsIn: z.string().max(100).trim(), // Human readable
}).strict().refine(
  (data) => {
    // callsToday should be <= maxPerDay
    return data.callsToday <= data.maxPerDay;
  },
  { message: "callsToday must be <= maxPerDay", path: ["callsToday"] }
).refine(
  (data) => {
    // callsThisSession should be <= maxPerSession if set
    if (data.maxPerSession !== undefined) {
      return data.callsThisSession <= data.maxPerSession;
    }
    return true;
  },
  { message: "callsThisSession must be <= maxPerSession", path: ["callsThisSession"] }
);

export type RateLimitStatus = z.infer<typeof RateLimitStatusSchema>;

// ============================================================================
// Stored Rate Limits Schema
// ============================================================================

/**
 * Rate limits stored in localStorage
 */
export const StoredRateLimitsSchema = z.object({
  date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/), // YYYY-MM-DD format
  counts: z.record(z.string().max(MAX_NAME_LENGTH), nonNegativeInt.max(100000)).refine(
    (counts) => Object.keys(counts).length <= 1000,
    { message: "Maximum 1000 rate limit keys" }
  ),
  lastReset: z.string().datetime(), // ISO timestamp
}).strict();

export type StoredRateLimits = z.infer<typeof StoredRateLimitsSchema>;

// ============================================================================
// Validation Helpers
// ============================================================================

export function validateStoredRateLimits(data: unknown): StoredRateLimits {
  return StoredRateLimitsSchema.parse(data);
}

export function safeValidateStoredRateLimits(data: unknown) {
  return StoredRateLimitsSchema.safeParse(data);
}

File: standalone/lattice-core/haskell/Lattice/Schemas/Settings/RateLimitsSchema.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : Lattice.Schemas.Settings.RateLimitsSchema
Description : Rate limits schema
Copyright   : (c) Lattice, 2026
License     : MIT

Validates rate limit data stored in localStorage/sessionStorage.

Source: lattice-core/lean/Lattice/Schemas/Settings/RateLimitsSchema.lean
-}

module Lattice.Schemas.Settings.RateLimitsSchema
  ( -- * Constants
    maxCallsPerDay
  , maxCallsPerSession
  , maxVramCostMB
  , maxRateLimitKeys
    -- * Rate Limit Config
  , RateLimitConfig(..)
  , validateRateLimitConfig
    -- * Rate Limit Status
  , RateLimitStatus(..)
  , validateRateLimitStatus
    -- * Stored Rate Limits
  , StoredRateLimits(..)
  , validateStoredRateLimits
  , safeValidateStoredRateLimits
  ) where

import GHC.Generics (Generic)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Maybe (Maybe(..))

import Lattice.Schemas.SharedValidation
  ( ValidationError, mkError, maxNameLength
  , validateNonEmptyString, validateString
  , validatePositiveInt, validateNonNegativeInt
  , validatePositiveFloat, validateDateTime, validateDate
  )

--------------------------------------------------------------------------------
-- Constants
--------------------------------------------------------------------------------

maxCallsPerDay :: Int
maxCallsPerDay = 100000

maxCallsPerSession :: Int
maxCallsPerSession = 10000

maxVramCostMB :: Double
maxVramCostMB = 1000000  -- 1TB

maxRateLimitKeys :: Int
maxRateLimitKeys = 1000

--------------------------------------------------------------------------------
-- Rate Limit Config
--------------------------------------------------------------------------------

-- | Rate limit configuration for a tool
data RateLimitConfig = RateLimitConfig
  { rlcToolName :: !Text
  , rlcMaxPerDay :: !Int
  , rlcMaxPerSession :: !(Maybe Int)
  , rlcVramCostMB :: !(Maybe Double)
  }
  deriving stock (Eq, Show, Generic)

-- | Validate RateLimitConfig
validateRateLimitConfig :: RateLimitConfig -> Either ValidationError RateLimitConfig
validateRateLimitConfig config = do
  _ <- validateNonEmptyString "toolName" maxNameLength (rlcToolName config)
  _ <- validatePositiveInt "maxPerDay" maxCallsPerDay (rlcMaxPerDay config)
  case rlcMaxPerSession config of
    Just maxSess -> do
      _ <- validatePositiveInt "maxPerSession" maxCallsPerSession maxSess
      pure ()
    Nothing -> pure ()
  case rlcVramCostMB config of
    Just cost -> do
      _ <- validatePositiveFloat "vramCostMB" maxVramCostMB cost
      pure ()
    Nothing -> pure ()
  Right config

--------------------------------------------------------------------------------
-- Rate Limit Status
--------------------------------------------------------------------------------

-- | Rate limit status for a tool
data RateLimitStatus = RateLimitStatus
  { rlsToolName :: !Text
  , rlsCallsToday :: !Int
  , rlsMaxPerDay :: !Int
  , rlsCallsThisSession :: !Int
  , rlsMaxPerSession :: !(Maybe Int)
  , rlsCanCall :: !Bool
  , rlsBlockedReason :: !(Maybe Text)
  , rlsResetsAt :: !Text  -- ISO timestamp
  , rlsResetsIn :: !Text  -- Human readable
  }
  deriving stock (Eq, Show, Generic)

-- | Validate status constraints
validateStatusConstraints :: RateLimitStatus -> Either ValidationError ()
validateStatusConstraints status
  | rlsCallsToday status > rlsMaxPerDay status =
      Left $ mkError "callsToday" "must be <= maxPerDay"
  | otherwise = case rlsMaxPerSession status of
      Just maxSess | rlsCallsThisSession status > maxSess ->
        Left $ mkError "callsThisSession" "must be <= maxPerSession"
      _ -> Right ()

-- | Validate RateLimitStatus
validateRateLimitStatus :: RateLimitStatus -> Either ValidationError RateLimitStatus
validateRateLimitStatus status = do
  _ <- validateNonEmptyString "toolName" maxNameLength (rlsToolName status)
  _ <- validateNonNegativeInt "callsToday" maxCallsPerDay (rlsCallsToday status)
  _ <- validatePositiveInt "maxPerDay" maxCallsPerDay (rlsMaxPerDay status)
  _ <- validateNonNegativeInt "callsThisSession" maxCallsPerSession (rlsCallsThisSession status)
  case rlsMaxPerSession status of
    Just maxSess -> do
      _ <- validatePositiveInt "maxPerSession" maxCallsPerSession maxSess
      pure ()
    Nothing -> pure ()
  case rlsBlockedReason status of
    Just reason -> do
      _ <- validateString "blockedReason" 500 reason
      pure ()
    Nothing -> pure ()
  _ <- validateDateTime "resetsAt" (rlsResetsAt status)
  _ <- validateString "resetsIn" 100 (rlsResetsIn status)
  validateStatusConstraints status
  Right status

--------------------------------------------------------------------------------
-- Stored Rate Limits
--------------------------------------------------------------------------------

-- | Rate limits stored in localStorage
data StoredRateLimits = StoredRateLimits
  { srlDate :: !Text  -- YYYY-MM-DD
  , srlCounts :: ![(Text, Int)]
  , srlLastReset :: !Text  -- ISO timestamp
  }
  deriving stock (Eq, Show, Generic)

-- | Validate StoredRateLimits
validateStoredRateLimits :: StoredRateLimits -> Either ValidationError StoredRateLimits
validateStoredRateLimits limits = do
  _ <- validateDate "date" (srlDate limits)
  _ <- validateDateTime "lastReset" (srlLastReset limits)
  if length (srlCounts limits) > maxRateLimitKeys
    then Left $ mkError "counts" $
      "must have at most " <> T.pack (show maxRateLimitKeys) <> " keys"
    else do
      mapM_ validateCount (srlCounts limits)
      Right limits
  where
    validateCount (key, count) = do
      _ <- validateString "counts.key" maxNameLength key
      _ <- validateNonNegativeInt "counts.value" maxCallsPerDay count
      pure ()

-- | Safe validation (returns Maybe)
safeValidateStoredRateLimits :: StoredRateLimits -> Maybe StoredRateLimits
safeValidateStoredRateLimits limits =
  case validateStoredRateLimits limits of
    Right l -> Just l
    Left _ -> Nothing
